mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala

```diff
@@ -0,0 +1,631 @@
+      val eachStats = data.mapPartitions { iter =>
+        val map = mutable.Map.empty[Long, (BV[Double], Double, BV[Double])]
+        iter.foreach { case (idx, point) =>
+          // calculate next index number
+          val childrenCenters = Array(2 * idx, 2 * idx + 1)
+              .filter(bcNewCenters.value.keySet.contains(_)).map(bcNewCenters.value(_)).toArray
+          if (childrenCenters.size >= 1) {
+            val closestIndex =
+              HierarchicalClustering.findClosestCenter(bcMetric.value)(childrenCenters)(point)
+            val nextIndex = 2 * idx + closestIndex
+
+            // get a map value or else get a sparse vector
+            val (sumBV, n, sumOfSquares) = map.get(nextIndex)
+                .getOrElse(BSV.zeros[Double](point.size), 0.0, BSV.zeros[Double](point.size))
+            map(nextIndex) = (sumBV + point, n + 1.0, sumOfSquares + (point :* point))
+          }
+        }
+        map.toIterator
+      }.reduceByKey { case ((sv1, n1, sumOfSquares1), (sv2, n2, sumOfSquares2)) =>
+        // sum the accumulation and the count in the all partition
+        (sv1 + sv2, n1 + n2, sumOfSquares1 + sumOfSquares2)
+      }.collect().toMap
+
+      // calculate the center of each cluster
+      newCenters = eachStats.map { case (idx, (sum, n, sumOfSquares)) => (idx, sum :/ n)}
+      bcNewCenters = sc.broadcast(newCenters)
+
+      // update summary of each cluster
+      stats = eachStats.toMap
+
+      variances = stats.map { case (idx, (sum, n, sumOfSquares)) =>
+        math.pow(sumOfSquares.toArray.sum, 1.0 / sumOfSquares.size)
+      }.sum
+      diffVariances = math.abs(oldVariances - variances) / oldVariances
+      oldVariances = variances
+      subIter += 1
+    }
+    stats
+  }
+
+  /**
+   * Gets the initial centers for bi-sect k-means
+   */
+  private[clustering]
+  def initChildrenCenter(clusters: Map[Long, BV[Double]]): Map[Long, BV[Double]] = {
+    val rand = new XORShiftRandom()
+    rand.setSeed(this.seed)
+
+    clusters.flatMap { case (idx, center) =>
+      val childrenIndexes = Array(2 * idx, 2 * idx + 1)
+      val relativeErrorCoefficient = 0.001
+      Array(
+        (2 * idx, center.map(elm => elm - (elm * relativeErrorCoefficient * rand.nextDouble()))),
+        (2 * idx + 1, center.map(elm => elm + (elm * relativeErrorCoefficient * rand.nextDouble())))
+      )
+    }.toMap
+  }
+
+  /**
+   * Builds a cluster tree from a Map of clusters
+   *
+   * @param treeMap divided clusters as a Map class
+   * @param rootIndex index you want to start
+   * @param numClusters the number of clusters you want
+   * @return a built cluster tree
+   */
+  private[clustering]
+  def buildTree(treeMap: Map[Long, ClusterTree],
+    rootIndex: Long,
+    numClusters: Int): Option[ClusterTree] = {
+
+    // if there is no index in the Map
+    if (!treeMap.contains(rootIndex)) return None
+
+    // build a cluster tree if the queue is empty or until the number of leaves clusters is enough
```

Comment:
grammar nit, change `until the number of leaves clusters` -> `until the number of leaf clusters`